docs/contribute/doc.md

```diff
@@ -0,0 +1,116 @@
+# Documentation style
+
+We are in the process of implementing new documentation requirements for mathlib. All future pull
+requests must meet the following standards.
+
+## Header comment
+
+Each mathlib file should start with a header comment with copyright information (see example below)
+followed by general documentation written using Markdown. Headers use atx-style headers (with hash
+signs, no underlying dash).
+
+The mandatory title of the file is a first level header. It is followed by a summary of the content
+of the file.
+
+The other sections, with second level headers are (in this order):
+* *Main definitions* (optional, can be covered in the summary)
+* *Main statements* (optional, can be covered in the summary)
+* *Notations* (ommited only if no notation is introduced in this file)
+* *Implementation notes* (description of important design decisions or interface features)
+* *References* (references to textbooks or papers, or Wikipedia pages)
+
+After references, we write "Tags:" followed by a list of keywords that could be useful when
+doing text search in mathlib to find where something is covered.
+
+The following code block is an example of a file header.
+
+```
+/-
```

Comment:
Lean natively supports module-level docstring with the `/-! .. -/` syntax - perhaps it'd be worth it to suggest using it here.

Reply:
Oh, nice, I saw this at one point but completely forgot. Unfortunately, they're only accessible from within that module -- if `file2.lean` imports `file1.lean`, the module docstrings in `file1` aren't accessible in `file2`. Also, these docstrings can't appear before `import`, so we'd have to separate this from the header or change the style there.

Given the restriction with importing, I'm not sure we gain much from using these. Do you know what the `option name` is in the output of `tactic.module_doc_strings`?

Reply:
The limitation on requiring imports to be the first token to appear at the parser stage is a bit strange, but a corresponding change in the style guide here doesn't seem to drastic to me.

There is a clear advantage to having docstrings instead of plain comments - they are a semantically meaningful part of the module (while comments are thrown away), which is (or should be..) serialized in olean output, as well as available to tactics. We can directly query Lean for them, and then export that into nice, human-readable documentation.

Regarding `meta constant module_doc_strings : tactic (list (option name × string))`, the short answer is that `(some n, doc)` means "declaration `n` is documented with `doc`", while `(none, doc)` means "the module has `doc` as a docstring". This whole thing is a bit strange and possibly not what was intended - I started working on some fix to that in leanprover-community/lean#54.

Reply:
How are you fixing it?

Reply:
Sorry, I wrote down the wrong statement. The (option name, string) convention seems reasonable - I'm fixing the doc output not working, hopefully without changing the interface.

Reply:
Ah, I was misinterpreting `module_doc_strings` to only return the module-level doc strings. Instead it gets all the doc strings in the module. But it's still true that there's no Lean API to retrieve module-level doc strings outside of the current module, right?

Reply:
As far as I know, that's right - currently, once a module has been built, its docstrings seem to disappear in the depths of the compiler. But this should be fixable.